client/state/themes/actions.js

```diff
@@ -454,3 +460,83 @@ export function clearThemeUpload( siteId ) {
 		siteId,
 	};
 }
+
+export function initiateThemeTransfer( siteId, file ) {
+	return dispatch => {
+		dispatch( {
+			type: THEME_TRANSFER_INITIATE_REQUEST,
+			siteId,
+		} );
+		return wpcom.undocumented().initiateTransfer( siteId, null, file, ( event ) => {
+			dispatch( {
+				type: THEME_TRANSFER_INITIATE_PROGRESS,
+				siteId,
+				loaded: event.loaded,
+				total: event.total,
+			} );
+		} )
+			.then( ( { transfer_id } ) => {
+				dispatch( {
+					type: THEME_TRANSFER_INITIATE_SUCCESS,
+					siteId,
+					transferId: transfer_id,
+				} );
+				themeTransferStatus( siteId, transfer_id )( dispatch );
+			} )
+			.catch( error => {
+				dispatch( {
+					type: THEME_TRANSFER_INITIATE_FAILURE,
+					siteId,
+					error,
+				} );
+			} );
+	};
+}
+
+export function themeTransferStatus( siteId, transferId, interval = 3000, timeout = 180000 ) {
+	const endTime = Date.now() + timeout;
+	return dispatch => {
+		const pollStatus = ( resolve, reject ) => {
+			if ( Date.now() > endTime ) {
+				// timed-out, stop polling
+				dispatch( {
+					type: THEME_TRANSFER_STATUS_FAILURE,
+					siteId,
+					transferId,
+					error: 'client timeout',
+				} );
+				return resolve();
```

Comment:
Yeah, I initially had a `reject()` here. But. The promise is only returned for testing purposes, so the production calling code won't be expecting to need a catch() call, meaning we would have an un-handled rejection on an API error. The important part is the `dispatch` of the error, I think.